File: 0283. Move Zeroes/0283.java
// time complexity: O(n^2) at worst case scenario
// space complexity: O(1)
class Solution {
    public void moveZeroes(int[] nums) {
        // it should be in-place algorithm
        for(int i=1; i<nums.length; i++){
            if(nums[i-1]==0){ // if previous is zero, find the next non-zero number
                int j=i;
                while (j<nums.length && nums[j]==0) j++;
                if(j<nums.length) swap(nums, i-1, j);
            }
        }
    }

    // a simple swap function
    public void swap(int[] nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 0283. Move Zeroes/0283.py
# time complexity O(n)
# space complexity O(1)
class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        j=0 # pointer to track position of the next non-zero element
        for i in range(0,len(nums),1):
            if nums[i] != 0:
                if i != j: # if the pointers at different indexes, swap.
                    nums[i], nums[j] = nums[j], nums[i] # swap func. in python
                j+=1
        